### PR TITLE
Finish touching nf_placement.py

### PR DESCRIPTION
**Pull request by janeyen**
Add comments, remove unused variables for nf_placement.py file

### CODE REVIEW

File: src/nf_heuristic.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
+from __future__ import print_function
```
Comment (jwangee) [inline src/nf_heuristic.py:1]: Please add several sentences to describe this file. Thanks!

File: src/nf_heuristic.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+        for prev in module_list[i+1].prev_nodes:
+            print("%s "% prev.nf_class)
+
+def parse_all_node(nfcp_parser):
+    nf_graph = convert_global_nf_graph(nfcp_parser.scanner)
+    modules = nf_graph.list_modules()
+    modules.sort(key=lambda l: (l.service_path_id, l.service_id))
+    return modules
+
+def init_allp4(module_list):
```
Comment (jwangee) [inline src/nf_heuristic.py:50]: In my impression, the following several functions are critical. It is helpful if you could add 1-2 sentenses saying what it does. Also, remember to remove unnecessary comments from the code.

File: src/nf_heuristic.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+        module_list[fix_index].offload = True
+        module_list[fix_index].nf_type = 1
+        option_nf = list(option_set)
+        module_list = next_placement_to_offload_p4(module_list)
+    else:
+        print("NO AVAILABLE PLACEMENT")
+
+    return module_list
+
+def highest_core_allocation_derived_from_LP(module_list):
```
Comment (jwangee) [inline src/nf_heuristic.py:139]: Sorry, this function name is a little bit confused to me.
Do you prefer core_alloc_highest_tput_derived_from_LP for a better readability? Or others if you prefer.

File: src/nf_heuristic.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        find_solution = True
+        chosen_pattern, core_alloc, expected_throughput = placeTool.optimize_pick(throughput_list)
+        chosen_module_list = placeTool.apply_pattern(module_list, chosen_pattern, core_alloc, bess_dict_para)
+    else:
+        find_solution = False
+#        chosen_module_list = module_list
+        expected_throughput = 0
+    return find_solution, chosen_module_list, expected_throughput
+
+def least_change(throughput_dict):
```
Comment (jwangee) [inline src/nf_heuristic.py:157]: In my impression, least_change, can_offload, and potential_better_placement are different heuristic algorithms mentioned in the paper. I'd prefer to clarify them a little bit with comments. Something like: this function implements 'xxx' heuristics, which is conservative / aggresive. Thanks.

File: src/optimization.py [hunk 5]
```diff
@@ -2,12 +2,12 @@
 import gurobipy as gp
 
 
-# bounds = ((min_f1, max_f1), ..., (min_fk, max_fk))
-# A f <= b
-# A is a matrix and b is a column vector
-# return (throughput_f1, ... throughput_fk) > (0, ..., 0) if feasible;
-# otherwise (0, ..., 0)
 def maximizeMarginalRate(bounds, A, b, noLog=True):
+    # bounds = ((min_f1, max_f1), ..., (min_fk, max_fk))
```
Comment (jwangee) [inline src/optimization.py:6]: Thanks. If it is possible, I'd suggestion adding the definition of A, b so that we know what is the input.

**Review by jwangee — COMMENTED** [review]
Thanks. I left a few nit comments. Otherwise, looks good.

**Comment by janeyen** [comment]
I haven't touched nf_heuristic.py and optimization.py actually. If the nf_placement file looks okay, could you merge this one? I will do another PR for other files

**Comment by jwangee** [comment]
Okay.

**Pull request merged**